### PR TITLE
Move samd21 to samr21

### PR DESCRIPTION
**Pull request by keestux**
Here  is a new PR to move samd21 to samr21 and to make room for adding samd21.

I promise to do my best to create / merge the code into common pieces that I will place in sam21_common. But I want this change to be merged so that the next steps can be small(er). Well, except for my board addition plus cpu/samd21.

### CODE REVIEW

**Comment by aabadie** [comment]
I'm not sure to be the right person to review this but I'm interested as I want to make RIOT run on an Arduino zero board (#5475).
@kYc0o or @PeterKietzmann, can you have a look ?

I'll make a comment inline.

File: doc/doxygen/riot.doxyfile [hunk 1]
```diff
@@ -818,13 +818,10 @@ EXCLUDE_SYMLINKS       = NO
 # exclude all test directories for example use the pattern */test/*
 
 EXCLUDE_PATTERNS       = */board/*/tools/* \
-                         */cpu/*/include/component/* \
```
Comment (aabadie) [inline doc/doxygen/riot.doxyfile]: why removing those three lines ?

Comment (keestux) [inline doc/doxygen/riot.doxyfile]: The only directory that matched this pattern was the one in cpu/samd21 (now cpu/samr21), same as the pio and instance directory, I moved them to a new subdirectory cpu/sam*/include/cmsis. I found that adding a new subdirectory made it more obvious that this is part of Atmel CMSIS.

Comment (keestux) [inline doc/doxygen/riot.doxyfile]: Adding Arduino Zero should be become easier now. But perhaps you can wait after I've added my SODAQ Autonomo. You can then look at how I did the board configuration.

**Comment by PeterKietzmann** [comment]
I will have a detailed look next week. Sorry for the delay, but we're currently in feature freeze anyway.

**Comment by kYc0o** [comment]
This PR will take quite longtime to be reviewed, since it had already a long discussion. For now what I can say is that I'm not sure a CPU samr21 can really be called CPU. I'm aware about the discussion on this topic, but I'm not well convinced...

**Comment by PeterKietzmann** [comment]
Then what is your proposal ? We already discussed a lot in other PRs (this is the third PR Kees opened) and I wrote a long explanation to the maintainer mailing list where I asked for objections (no one objected). I would really like to have a solution for this issue.

**Comment by kYc0o** [comment]
I agree on the fact these CPUs need a common place where to put common code, since they share several features and architectural design. However, the special case of samr21 can be the only "CPU" with different characteristics. We can compare it for instance with the cc2538, which is also a cortex based MCU with a radio transceiver. The big difference here and why it is placed on a CPU folder is because the radio features are embedded in the same memory space, thus it really embeds the radio in the same chip. The case of samr21 is different, because it explicity separates a samd21 CPU with an attached at86rf2xx radio via SPI, thus it is an SoC on which two MCU, separately manufactured, have been put together. See [here](

File: doc/doxygen/riot.doxyfile (hunk 1 above)
Comment (aabadie) [inline doc/doxygen/riot.doxyfile]: > The only directory that matched this pattern was the one in cpu/samd21 (now cpu/samr21), same as the pio and instance directory, I moved them to a new subdirectory cpu/sam*/include/cmsis

Got it. Indeed, I missed that it was moved to a subdirectory of cmsis.

**Comment by kYc0o** [comment]
Sorry for closing and reopening, bad manipulation.

File: doc/doxygen/riot.doxyfile (hunk 1 above)
Comment (aabadie) [inline doc/doxygen/riot.doxyfile]: > Adding Arduino Zero should be become easier now. But perhaps you can wait after I've added my SODAQ Autonomo. You can then look at how I did the board configuration.

That will be great. Do you make the pads configurable somewhere with a macro ?

**Comment by keestux** [comment]
@kYc0o Please stop calling it a "SAMD21 CPU". There is no such thing as a SAMD21 CPU. All these (samr21, samd21, saml21) are a Cortex M0. All with the same instruction set. Only with different peripherals. Have a look at the 12 "CPU"s that Atmel has. SAMD21 is just one of them.

It would be really bad to to have a long review period. All that time I can't finish adding my SODAQ Autonomo. Then I have another board (SODAQ ONE). There is also interest to add Arduino Zero. That can't be done without accepting this PR first.

We went through several alternatives. None can give a quick solution that makes everybody happy. This #5645 is just a step forwards. I'll continue to work on it and move common code to cpu/sam21_common (bad name BTW).

**Comment by kYc0o** [comment]
> @kYc0o Please stop calling it a "SAMD21 CPU". There is no such thing as a SAMD21 CPU. All these (samr21, samd21, saml21) are a Cortex M0. All with the same instruction set. Only with different peripherals. Have a look at the 12 "CPU"s that Atmel has. SAMD21 is just one of them.

Here I'm talking about our concept of CPU, on which we use such name for a MCU, aiming to separate MCU and internal peripherals code from other entities present on an IoT device (radio, sensors, mass storage, etc.). Thus, either we continue using such nomenclature or someone makes a proposal to call it differently, hoping this would avoid all this misunderstanding.

> It would be really bad to to have a long review period. All that time I can't finish adding my SODAQ Autonomo. Then I have another board (SODAQ ONE). There is also interest to add Arduino Zero. That can't be done without accepting this PR first.

This long review period is not what I want obviously, I want rather to be realistic on how much time the people working on this kind of MCU can put into reviewing this, since it's quite delicate as some of our reference platforms use a lot this code (samr21-xpro, etc).
Me, as much ash you, I'm very interested on having support for your platform, and of course the new ones and the Arduino Zero, I know this is your third attempt but we really need to be clear on how we will proceed with the code factorisation, in order to make it more readable, standard and reliable.

> We went through several alternatives. None can give a quick solution that makes everybody happy. This #5645 is just a step forwards. I'll continue to work on it and move common code to cpu/sam21_common (bad name BTW).

I'm aware of the several alternatives, but as I have stated before, this requires a delicate work which requires time, this MCU is very flexible and growing in popularity, thus if our code is good and well reviewed we would support platforms based on this MCU easily and faster.
If you think `cpu/sam21_common` is a bad name, let's discuss it, I think talking and giving our point of view while being convincing with good arguments is the way to get this work merged as soon as possible, I'm really looking forward to it.

Thanks again for your good work and be sure that I'm as much interested as you on having your platform on our upstream RIOT repository.

**Comment by keestux** [comment]
> Here I'm talking about our concept of CPU, on which we use such name for a MCU, aiming to separate MCU and internal peripherals code from other entities present on an IoT device (radio, sensors, mass storage, etc.). Thus, either we continue using such nomenclature or someone makes a proposal to call it differently, hoping this would avoid all this misunderstanding.

In the case of Atmel, they have 12 CPU's (or MCU's or whatever we name them) that are very much alike. All are Cortex M0. All have a similar set of peripherals (SERCOM, timers, etc). Atmel has supplied a set of include files (CMSIS) in their ASF. Until now RIOT has added two "cpu"s from that set. One in cpu/samd21 and one in cpu/saml21. But the one in cpu/samd21 is in fact a SAMR21. The include files in cpu/samd21/include came from from Atmel but were renamed to samd21. (( This, BTW, is my biggest hurdle. ))

What I want say is, we already have a situation.

> This long review period is not what I want obviously, I want rather to be realistic on how much time the people working on this kind of MCU can put into reviewing this, since it's quite delicate as some of our reference platforms use a lot this code (samr21-xpro, etc).

I'd say, try out my PR and you'll see that it continues to work. Under the surface it is merely a name change samd21-to-samr21.

**Comment by keestux** [comment]
> Me, as much ash you, I'm very interested on having support for your platform, and of course the new ones and the Arduino Zero, I know this is your third attempt but we really need to be clear on how we will proceed with the code factorisation, in order to make it more readable, standard and reliable.

Oh, believe me. I am all for readable code. And I hope that my new code for the SERCOMs (UART, I2C, SPI) improves readability and ease of configuration. This new code will first show up in cpu/samd21 or cpu/sam21_common and then the samr21 or saml21 branches can be gradually changed to it it too, making them leaner.

**Comment by kYc0o** [comment]
> In the case of Atmel, they have 12 CPU's (or MCU's or whatever we name them) that are very much alike. All are Cortex M0. All have a similar set of peripherals (SERCOM, timers, etc). Atmel has supplied a set of include files (CMSIS) in their ASF. Until now RIOT has added two "cpu"s from that set. One in cpu/samd21 and one in cpu/saml21. But the one in cpu/samd21 is in fact a SAMR21. The include files in cpu/samd21/include came from from Atmel but were renamed to samd21. (( This, BTW, is my biggest hurdle. ))

OK I got your point but I need first to be sure why you insist on put the SAMR21 in the CPU forlder, assuming that we call them CPUs but they are MCUs. Looking at the Atmel MCU SAMx offer [here](http://www.atmel.com/products/microcontrollers/arm/default.aspx), I cannot see any SAM R Cortex-M0 based MCU. However the SAM D and SAM L are there.

> I'd say, try out my PR and you'll see that it continues to work. Under the surface it is merely a name change samd21-to-samr21.

On my understanding the SAMR21 SoC is composed by a SAMD21 MCU + AT86RF233 radio transceiver, thus I don't see a reason to rename the current `cpu/samd21` to `cpu/samr21`. On the contrary, I think this could lead to a misunderstanding on which CPU code should be used, since other boards like the Arduino Zero are based on a samd21 MCU. Thus using code from `cpu/samr21` wouldn't be logic. However, I can easily understand why a samr21 based board can use the code from `cpu/samd21` since it embeds a samd21 MCU, even if there's no explicit reference to a samr21 folder, wherever it would be located.

> Oh, believe me. I am all for readable code. And I hope that my new code for the SERCOMs (UART, I2C, SPI) improves readability and ease of configuration. This new code will first show up in cpu/samd21 or cpu/sam21_common and then the samr21 or saml21 branches can be gradually changed to it it too, making them leaner.

I completely agree on this point. If there is code that can be shared by several SAMx21 MCU it needs to be factorised. I'd propose to put it in a folder called `cpu/samx21_common`, assuming that our current supported saml21 MCU can also use code from there (and that Atmel won't release a SAM X MCU).

**Comment by PeterKietzmann** [comment]
Obviously we have a somehow tricky situation here to adapt Atmels product series in RIOT conventions. So whatever solution it would be at the end, it will be a compromise. We should aim for a solution that (i) is easy to understand and "somehow " logic (ii) does not introduce unnecessary subdirectory depth (iii) outsources as many code as possible to a common folder and (iv) makes use of the appropriate CMSIS header file. Now lets imagine our CPU folder would be named "Chip, Engine, Board-heart", the explicit conflict CPU vs MCU vs SoC etc won't be present any more. In that case I find it quite logical to search for a 
- "samr21" when I'm developing on an Atmel samr21-xpro board
- "saml21" when I'm developing on an Atmel saml21-xpro board
- "samd21" when I'm developing on an Arduino zero board

@kYc0o did I already convince you ;-)?  After our private discussion today, I looked at older discussions and previous PRs and I'm sill in favour of the solution proposed in here. Do you have an other concrete proposal?

**Comment by keestux** [comment]
> OK I got your point but I need first to be sure why you insist on put the SAMR21 in the CPU forlder, assuming that we call them CPUs but they are MCUs. Looking at the Atmel MCU SAMx offer here, I cannot see any SAM R Cortex-M0 based MCU. However the SAM D and SAM L are there.

I know. But samr21 is not listed among the SAM D series. And, in the documentation of the SAMR21 there is no mention at all that it is a SAMD21.

And I'm not insisting to have a cpu/samr21. I did (do) have a PR where cpu/samd21 is reorganized to handle samr21, samd21, and possibly others. But this is a more intrusive change. There are several places in the code where there are hard coded numbers (mux, pad, etc). I won't be surprised if that PR takes longer to be accepted.

> On my understanding the SAMR21 SoC is composed by a SAMD21 MCU + AT86RF233 radio transceiver, thus I don't see a reason to rename the current cpu/samd21 to cpu/samr21. On the contrary, I think this could lead to a misunderstanding on which CPU code should be used, since other boards like the Arduino Zero are based on a samd21 MCU. Thus using code from cpu/samr21 wouldn't be logic. However, I can easily understand why a samr21 based board can use the code from cpu/samd21 since it embeds a samd21 MCU, even if there's no explicit reference to a samr21 folder, wherever it would be located.

Point me at the documentation where Atmel says that it is a SAMD21. I haven't found it.

It's your assumption that Arduino Zero would use cpu/samr21, but why? There will be a cpu/samd21. Arduino Zero has a SAMD21G18A and SODAQ Autonomo has a SAMD21J18A. These are samd21. And they need a different include file. I know this because I have made the Arduino board file for SODAQ Autonomo.

> I completely agree on this point. If there is code that can be shared by several SAMx21 MCU it needs to be factorised. I'd propose to put it in a folder called cpu/samx21_common, assuming that our current supported saml21 MCU can also use code from there (and that Atmel won't release a SAM X MCU).

Notice that the latest is SAML22. And even in the past there were SAMD20, SAMD11, etc. That is why I said that sam21_common is misleading.

**Comment by kaspar030** [comment]
> In that case I find it quite logical to search for a
> 
>   * "samr21" when I'm developing on an Atmel samr21-xpro board
>   * "saml21" when I'm developing on an Atmel saml21-xpro board
>   * "samd21" when I'm developing on an Arduino zero board

Well, that distinction should be made in boards/.

Check this figure (page 9, 3.2 SAM R21 Interconnection):
http://www.atmel.com/Images/Atmel-42223%E2%80%93SAM-R21_Datasheet.pdf

The samr21 is a samd21 MCU and a at86fr233, together on the same die.

The MCU side will behave as all the other variants of the samd21, in the
code, (apart from minor differences). That's what counts: we can reuse
the same code.
After all refactoring, the cpu/samd21/ and cpu/samr21/ folders should be
mostly identical, apart from minor ifdefs for different periph
configuration, making the cpu/samr21 folder obsolete.

CMSIS complicates this, as there are actual header files with samr21 in
it. But they can easily reside within samd21/include/cmsis/samr21g18a,
or whatever the original name was.

The saml21 is very similar regarding the IO units, only the clock
subsystem is different. But that means all the setup functions for
periph will be different, together with any lpm code. Those differences
were enough to create cpu/saml21. Doesn't mean that saml21 and samd21
cannot share more code. Depending on the amount of differences, it
either makes sense to unify cpu/samd21 and cpu/saml21 into something
like cpu/sam21 or cpu/samx21, or factor out shared code into
cpu/sam21_common, or whatnot.

**Comment by kaspar030** [comment]
> Point me at the documentation where Atmel says that it is a SAMD21. I haven't found it.

http://www.atmel.com/Images/Atmel-42223%E2%80%93SAM-R21_Datasheet.pdf

Page 9, 3.2 SAM R21 Interconnection, within the figure.

**Comment by kaspar030** [comment]
I think the main issue here is that the current samd21/ is specific to samr21 in too many places, making it hard to add other samd21/ variants.

**Comment by gebart** [comment]
I don't understand why this seem to stir up so much feelings from everyone involved, it's just a bunch of vendor header files, we didn't even write them!

Anyway, to try to find a solution: For easier maintenance, I suggest we go with the vendor (upstream) classification of the CPUs. I've looked through the headers from Atmel Software Framework (ASF) 3.31 (available [here](http://www.atmel.com/tools/AVRSOFTWAREFRAMEWORK.aspx)) and the differences between the headers are tiny but enough that there needs to be two separate directories, in my opinion. Merging the two would be possible with added preprocessor macros, but it would require us to maintain our modifications in the future, something that I am not looking forward to.

ASF has the following directory structure (roughly, there are some intermediate layers which I have omitted because of irrelevance)

  - sam
    - sam3n
    - sam3s
    - sam3s8
    - sam3u
    - sam3x
    - sam4c
    - sam4cm
    - sam4cm32
    - sam4cp
    - sam4e
    - sam4l
    - sam4n
    - sam4s
    - same70
    - samg
    - sams70
    - samv70
    - samv71
  - sam0
    - samb11
    - samc20
    - samc21
    - samd09
    - samd10
    - samd11
    - samd20
    - samd21
    - samda1
    - saml21
    - saml22
    - samr21

each of the CPU directories above have one top level header file per specific chip, such as `samd21g16a.h`, another chip-specific file in `pio/samd21g16a.h`, and then a bunch of files that are common for the family in the sub-directories `instance` and `component`. The files currently in RIOT in the cpu/samd21/include directory are from ASF and are taken from sam0/samr21 and specifically samr21g18a.

**I support this PR in the basic intent with the following motivations and considerations:**

rename samd21 -> samr21, The headers are very similar between samr21, samd21, but not identical. There are a few registers which are not defined for samr21 but is defined for samd21, and vice versa. Therefore, like I wrote above, if we want to minimize our own maintenance burden we should try to replicate the vendor classification. This goes for the periph drivers too, since they are shared between all sam0 in ASF, we should be able to do the same as well!

Furthermore, I suggest that we should rename sam21_common -> sam0_common. Motivation: since the drivers in ASF are shared between those chips, even between samx11, samx21, samx22, it makes sense that we try to make common periph driver implementations based on the same classification.

Each CPU family (samd21, samr21, saml21 etc) needs to have its own interrupt vector definition, since that depends on the hardware components. `periph_cpu.h` should be shared in sam0_common since it only configures the periph drivers' internal stuff, such as the GPIO_DEV macro etc (look at kinetis_common for another example of shared periph drivers). Also, lpm_arch.c should probably go in sam0_common.

**Long term:**
Attempt to migrate periph drivers from samd21+samr21 to sam0_common. 

Earlier when discussing the Atmel sam0 devices it has been mentioned that it is not feasible to migrate all drivers to sam21_common, but after a brief scan of the ASF sources I tend to disagree; There are only a few places in each driver which are using preprocessor `#if` based on hardware features, such as hardware support for SPI CS (there are many more preprocessor conditions in ASF that are based on software configuration, such as enabling SPI slave mode etc.)

**Comment by gebart** [comment]
I forgot to mention: Any new periph implementations should go in the sam0_common dir, not in a CPU specific dir. This way might make a PR take a little longer to be completed, but it will save us effort in the long run, and we will get all features on all CPUs which have the hardware components for it.

**Comment by keestux** [comment]
@kaspar030 OK, you convinced me.
(To my defence, it was pretty well hidden. And the samd21 text can only be found twice in the 1200 pages and nowhere on their website.)

So what do you suggest? I went through many hoops. Do you have another one? :-)

> CMSIS complicates this, as there are actual header files with samr21 in it. But they can easily reside within samd21/include/cmsis/samr21g18a, or whatever the original name was.

Ideally I want to make as few changes to the (original!) CMSIS as possible. The current CMSIS files (in `cpu/samd21`) were mistreated (renamed, modified). Let's try to avoid that.

One of my PR's (I begin to loose track) I suggested to have the CMSIS files in `cpu/samd21/include/cmsis` and to copy the necessary CMSIS files verbatim (samd21, samr21). And then select the correct variant in the board Makefile.include like this (for me):

    export CFLAGS += -D__SAMD21J18A__

But, still, if we do it like this, then I have a new challenge. To get my board supported the code must change first. I don't have a samr21-xpro, nor am I interested in that board. The code changes must happen before I can add my board (or Arduino Zero). So now, I'll be dependent on the willingness of others to test my code.

**Comment by PeterKietzmann** [comment]
For the record: #5556 is "the other" pull request proposing a solution where samr21 is covered below samd21. @kaspar030, as Joakim explained, the solution proposed in here must not prohibit to reuse a lot of code, except the vendor headers. But I think we all agree that we don't want to maintain those, so re-usability is not the point here. @gebart's proposal describes what I had in mind, in a more detailed way. Did he convince you?

I still prefer #5645 but actually I intend to go for a solution where others have stronger opinions than me. So again, after all discussions, what is your opinion on #5556 vs #5645 @kaspar030, @kYc0o?

@keestux whatever decision we make, I am willing to test. No worries :-).

**Comment by keestux** [comment]
Hey,

I am inclined to withdraw this PR.

Why?
* right now I can hardly make any progress to get my board up and running and accepted in RIOT (which is my only drive to do all this)
* I'm trying yet another alternative set of changes (hoping to make easier to get accepted)
* we're in feature freeze mode, how long should I wait?

Meanwhile, I am pushing new branches to https://github.com/keestux/RIOT

My current, new approach is:

* start with up to date master
* move the current samd21 CMSIS include files to cpu/sam21_common/include/cmsis (branch move-cmsis-to-sam21_common-2)
* update the CMSIS for samr21 to ASF 3.21 (branch update-samr21-asf-2)
* add the CMSIS for samd21, ASF 3.30 (branch add-cmsis-samd21)

I was able to move i2c.c to cpu/sam21_common (branch move-i2c-to-sam21_common-2).

Unfortunately the saml21 and samd21 implementations of the other periph modules are very different. Why? This is going to be really hard for me to move things to sam21_common.

**Comment by PeterKietzmann** [comment]
@keestux I'm really sorry. Obviously my proposal wasn't adopted by the community. @kaspar030, @kYc0o can you please give some feedback as you were against the approach presented here? Please don't withdraw this PR as long as we have no consensus about how to proceed.

I'm a bit confused why you  start to move cmsis headers into the common folder (branches mentioned above). Whose approach was that? However, splitting changes into "small" PRs is always a good idea. 

Regarding the feature freeze I hope we will finish within this week but it is hard to estimate because of our lack of (wo)manpower.

**Comment by keestux** [comment]
@PeterKietzmann Ideally we should have central place for the CMSIS of all sam0 (samd21, saml21, etc). And sam21_common is the best choice right now. And also, doing it this way I can avoid the discussion of having cpu/samr21 besides cpu/samd21. (At least, until I have created enough momentum to bang the drum again.)

Another advantage of the new approach is that more (git) history will survive.

**Comment by PeterKietzmann** [comment]
I am willing to review and test any of your future changes but currently I don't see myself as the one to decide how sort these CPUs/MCUs/SoCs/howevers.

**Comment by kaspar030** [comment]
> Unfortunately the saml21 and samd21 implementations of the other periph modules are very different. Why?

Back when I did the saml21 port, the samd21 implementations where ... shitty. I tried to re-use them, but because the clock initialization was very different, I assumed the rest would be, too. So I mostly wrote them from (mostly) scratch for saml21. I since then @haukepetersen rewrote most of the samd21 code.

**Comment by keestux** [comment]
@PeterKietzmann Basically I need someone with a samr21-xpro to test my branch
 (https://github.com/keestux/RIOT/tree/refactor-samd21-spi)

* UART
* I2C, and
* SPI

What are the possibilities?

**Comment by PeterKietzmann** [comment]
I am someone with a samr21-xpro and the possibility would be to test your stuff :-). If you just want to intermediately check your code, send me a private mail what should be tested. If it's already a contribution wish I would like to have consensus about the structuring-issue and systematically test everything based on a PR.

**Comment by keestux** [comment]
In a private conversation @PeterKietzmann  tried out the new branch ("refactor-samd21-spi") on a samr21-xpro. The result is that everything is still working on samr21-xpro.

So, I want to withdraw this PR, and send a new. Or, I can force-push and overwrite the current PR. Which do you prefer?

**Comment by PeterKietzmann** [comment]
I propose to open a new PR and close this one. 1. it is a cleaner distinction between both approaches and 2. it makes the topic more dramatic (thus, hopefully higher priorised) because it will be your 4th PR.

**Pull request closed without merging**